Regular expression concept with exercise (#2159)
exercism/go: concepts/regular-expressions/about.md

# About 

Package [regexp][package-regexp] offers support for regular expressions in Go.

## Syntax 

The [syntax][regexp-syntax] of the regular expressions accepted is the same general syntax used by Perl, Python, and other languages. 

> All characters are UTF-8-encoded code points. 

Author: What I was really trying to say was that both the patterns and the input texts are interpreted as sequences of UTF-8-encoded code points. I have changed this sentence accordingly.  Is that any better?

> Following `utf8.DecodeRune`, each byte of an invalid UTF-8 sequence is treated as if it encoded `utf8.RuneError (U+FFFD)`.

Author: Agreed

> It is convenient to write regular expressions as [raw string literals][raw-string-literals] enclosed by backticks.

Reviewer: ```suggestion
It is convenient to write regular expressions as [raw string literals][raw-string-literals] enclosed by backticks (\`).
```

> It is convenient to write regular expressions as [raw string literals][raw-string-literals] enclosed by backticks.

Reviewer: ```suggestion
It is convenient to write regular expressions as [raw string literals][raw-string-literals] enclosed by backticks (\`).
```

> This avoids having to quote the backslashes.

Author: Done

## Type `RegExp`

The package defines a type `Regexp` for compiled regular expressions.
Function `regexp.Compile` compiles a string into a regular expression.
The function returns `nil` and an error if compilation failed:

```go
re, err := regexp.Compile(`(a|b)+`)
fmt.Println(re, err) // => (a|b)+ <nil>
re, err = regexp.Compile(`a|b)+`)
fmt.Println(re, err) // => <nil> error parsing regexp: unexpected ): `a|b)+`
```

Function `MustCompile` is like `Compile` but panics if the expression cannot be parsed.

> It simplifies safe initialization of global variables holding compiled regular expressions.

Reviewer: I think we can expand on this a bit more here and compare the two. For instance, say something along the lines of `MustCompile` being easier to use because we don't need to handle the error, but also mention that it must be used with extreme caution and only when we know for sure the pattern does compile, otherwise the program will panic.

Author: I have moved MustCompile to about.md and expanded on it there.

> It simplifies safe initialization of global variables holding compiled regular expressions.

Reviewer: I think we can expand on this a bit more here and compare the two. For instance, say something along the lines of `MustCompile` being easier to use because we don't need to handle the error, but also mention that it must be used with extreme caution and only when we know for sure the pattern does compile, otherwise the program will panic.

  
There are 16 methods of `Regexp` that match a regular expression and identify the matched text.
Their names are matched by this regular expression:

```text
Find(All)?(String)?(Submatch)?(Index)?
```

* If `All` is present, the routine matches successive non-overlapping matches of the entire expressions.
* If `String` is present, the argument is a string; otherwise it is a slice of bytes; return values are adjusted as appropriate. 
* If `Submatch` is present, the return value is a slice identifying the successive submatches of the expression.

> * If `Index` is present, matches and submatches are identified by byte index pairs within the input string.

Reviewer: I like this bit a lot!

Author: Yes, it is cute describing regexp function syntax using regular expressions :)
(It's taken from the regexp package doc)

> * If `Index` is present, matches and submatches are identified by byte index pairs within the input string.

Reviewer: I like this bit a lot!

# About 

Package [regexp][package-regexp] offers support for regular expressions in Go.

## Syntax 

The [syntax][regexp-syntax] of the regular expressions accepted is the same general syntax used by Perl, Python, and other languages. 

Both the search patterns and the input texts are interpreted as UTF-8.  

When using backticks ``(`)`` to make strings, backslashes `(\)`  don't have any special
meaning, and don't mark the beginning of special characters like tabs `\t` or
newlines `\`n:

```go
"\t\n" // regular string literal with 2 characters: a tab and a newline
`\t\n`// raw string literal with 4 characters: two backslashes, a 't', and an 'n'
```

Because of this, using backticks is desirable to make regular expressions,
because it means we don't need to escape backslashes:

```go
"\\" // string with a single backslash
`\\` // string with 2 backslashes
```

> ## Type `RegExp`

Reviewer: ```suggestion
## Compiling patterns - `RegExp` type
```

> The package defines a type `Regexp` for compiled regular expressions.
> Function `regexp.Compile` compiles a string into a regular expression.
> The function returns `nil` and an error if compilation failed:

Reviewer: ```suggestion
To use a regular expression, we first must compile the string pattern.
Compilation here means taking the string pattern of the regular expression and converting it into an internal representation that is easier to work with.
We only need to compile each pattern once, after that we can use the compiled version of the regular expression many times.
The type `regexp.Regexp` represents a compiled regular expressions.
We can compile a string pattern into a `regexp.Regexp` using the function  `regexp.Compile`.
This function returns `nil` and an error if compilation failed:
```

I think this clarification is important. I remember when learning regexes, one of the things that most confused me was the need for a compilation step, so I think we should try to address this. 
Another thing I see people new to regexes struggle with is that they may accpet need to compile their regex, but then they think they need to compile it every time they need to use it.

```go
re, err := regexp.Compile(`(a|b)+`)
fmt.Println(re, err) // => (a|b)+ <nil>
re, err = regexp.Compile(`a|b)+`)
fmt.Println(re, err) // => <nil> error parsing regexp: unexpected ): `a|b)+`
```

Function `MustCompile` is a convenient alternative to `Compile`: 

```go 
re = regexp.MustCompile(`[a-z]+\d*`)
```

Using this function, there is no need to handle an error. 

~~~~exercism/caution
 `MustCompile` should only be used when we know for sure the pattern does compile, as otherwise the program will panic.
 ~~~~

>    
> There are 16 methods of `Regexp` that match a regular expression and identify the matched text.

Reviewer: ```suggestion
 
 ## Regular expression methods
 
There are 16 methods of `Regexp` that match a regular expression and identify the matched text.
```

Their names are matched by this regular expression:

```text
Find(All)?(String)?(Submatch)?(Index)?
```

* If `All` is present, the routine matches successive non-overlapping matches of the entire expressions.
* If `String` is present, the argument is a string; otherwise it is a slice of bytes; return values are adjusted as appropriate. 
* If `Submatch` is present, the return value is a slice identifying the successive submatches of the expression.
* If `Index` is present, matches and submatches are identified by byte index pairs within the input string.

There are also methods for:

* replacing matches of regular expressions with replacement strings and
* splitting of strings separated by regular expressions.

All-in-all, the `regexp` package defines more than 40 functions and methods.
We will demonstrate the use of a few methods below.
Please see the [API documentation][package-regexp] for details of these and other functions.
 

> ## Examples 

Reviewer: ```suggestion
### `MatchString` Examples 
```

> Method `MatchString` reports whether a strings contains any match of a regular expression.

Reviewer: ```suggestion
Method `MatchString` reports whether a string contains any match of a regular expression.
```

```go
re = regexp.MustCompile(`[a-z]+\d*`)
b = re.MatchString("[a12]")       // => true
b = re.MatchString("12abc34(ef)") // => true
b = re.MatchString(" abc!")       // => true
b = re.MatchString("123 456")     // => false    
```

>  
> Method `FindString` returns a string holding the text of the leftmost match of the regular expression.

Reviewer: ```suggestion

### `FindString` Examples 

Method `FindString` returns a string holding the text of the leftmost match of the regular expression.
```

Adding a header.

```go
re = regexp.MustCompile(`[a-z]+\d*`)
s = re.FindString("[a12]")       // => "a12"
s = re.FindString("12abc34(ef)") // => "abc34"
s = re.FindString(" abc!")       // => "abc"
s = re.FindString("123 456")     // => ""
```

> 
> Method `FindStringSubmatch` returns a slice of strings holding the text of the leftmost match of the regular expression and the matches, if any, of its subexpressions.

Reviewer: ```suggestion

### `FindStringSubmatch` Examples

Method `FindStringSubmatch` returns a slice of strings holding the text of the leftmost match of the regular expression and the matches, if any, of its subexpressions.
```

Adding a header.

This can be used to identify the strings matching capturing groups.
A return value of `nil` indicates no match.

```go 
re = regexp.MustCompile(`[a-z]+(\d*)`)
sl = re.FindStringSubmatch("[a12]")       // => []string{"a12","12"}
sl = re.FindStringSubmatch("12abc34(ef)") // => []string{"abc34","34"}
sl = re.FindStringSubmatch(" abc!")       // => []string{"abc",""}
sl = re.FindStringSubmatch("123 456")     // => <nil>
```

> 
> Method `re.ReplaceAllString(src,repl)` returns a copy of `src`, replacing matches of the regular expression `re` with the replacement string `repl`.

Reviewer: ```suggestion

### `ReplaceAllString` Examples

Method `re.ReplaceAllString(src,repl)` returns a copy of `src`, replacing matches of the regular expression `re` with the replacement string `repl`.
```

Adding a header.

```go
re = regexp.MustCompile(`[a-z]+\d*`)
s = re.ReplaceAllString("[a12]", "X")       // => "[X]"
s = re.ReplaceAllString("12abc34(ef)", "X") // => "12X(X)"
s = re.ReplaceAllString(" abc!", "X")       // => " X!"
s = re.ReplaceAllString("123 456", "X")     // => "123 456"
```

>   
> Method `re.Split(s,n)` slices a text `s` into substrings separated by the expression and returns a slice of the substrings between those expression matches.

Reviewer: ```suggestion
 
 ### `Split` Examples
 
Method `re.Split(s,n)` slices a text `s` into substrings separated by the expression and returns a slice of the substrings between those expression matches.
```

Adding a header.
